Turn off debuginfo for build dependencies to improve compile times

lqd opened the pull request:
This PR is a draft towards the possible state of build-override defaults described in #10481, in order to improve compile times for build dependencies (mostly proc-macros and their dependencies benefit from this):
- debuginfo is turned off by default (stripping and incremental are less impactful and can be done as improvements in future PRs, since they also require more analysis and work) to match [this comment](https://github.com/rust-lang/cargo/issues/10481#issuecomment-1068467069) and [zulip message](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Defaults.20for.20faster.20compilation.20of.20.22for.20host.22.20units/near/275717621)
- the new defaults are documented, and explain how to turn it back on when needed
- when a build script fails, and backtraces are requested, a custom note mentions how to improve backtraces and links to the documentation

Opening as draft for feedback and guidance:
- [x] on the build script error message itself (it tries to only show when applicable, when backtraces are opted into, but I wonder if that interferes with how diagnostics are usually buffered). **Update**: the current approach and message [look acceptable](https://github.com/rust-lang/cargo/pull/10493#issuecomment-1074345394).
- [x] a bunch of tests rely on the fact that dependencies and build dependencies will not be rebuilt under the current defaults: now that the `dev` profile defaults and `dev.build-override` differ on whether debuginfo is turned on, these dependencies will be built twice. To leave the tests as is, either debuginfo could be turned off, or manually set to 2, which I've done in the "tmp: update tests relying on dev and build-deps reuse" commit. I am not sure it's the expected way to do this ? **Update**: the test changes [look acceptable](https://github.com/rust-lang/cargo/pull/10493#issuecomment-1074345394) so I've turned the temporary commit into a permanent one.
- [x] I need some help for `-Zscrape-examples`: the feature seems to depend implicitly on a similar build reuse, with a mapping from for-host dependencies to regular dependencies. I'm not sure that this works correctly in all situations: it seems focused on feature flags, but reuse can also be different depending on other flags. In our case, debuginfo is now different, making some of the mapping memoization panic, causing the tests to fail. So I've temporarily disabled them in the "tmp: disable -Zscrape-examples unit tests" commit, and would need help to know what to do there. My expectation is that these panics would already happen today if someone manually opted into different debuginfo level in their `build-override` (or panic method, etc) but have not tested it (it seems sensible in a context where only the default settings have changed to a value users can already set). **Update**: This issue is now tracked in #10500. We've discussed it with @willcrichton, and I've incorporated a fix and tests in #10524.

---
**Edit**: with the updates above, this is ready to take out of draft.

The discussion issue #10481 also mentions the possible small `strip` addition that this PR doesn't do (and that could be done in the future), but since it's not a huge improvement, I'd say this closes #10481.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @ehuss (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

bjorn3 commented on src/cargo/core/profiles.rs:
```diff
@@ -418,8 +418,11 @@ impl ProfileMaker {
+            // less work.
             profile.opt_level = InternedString::new("0");
             profile.codegen_units = None;
+            profile.debuginfo = None;
```
Should this also add `strip = "debuginfo"` on all platforms but macOS?

lqd commented:
Thanks Josh, I'll do so ASAP.

(I've only now noticed I didn't open this PR as a draft GH PR as I intended to)

lqd commented:
Removed the temporary commits about:
- fixing the tests relying on reuse, as josh said it looked OK
- ignoring the scrape examples tests surfacing an existing problem, as I've opened a dedicated issue for that and it's not related to this PR

CI will fail because of #10500, and I'd assume it's kind of blocked on a resolution for that, otherwise `-Z rustdoc-scrape-examples` will likely stop working altogether.

lqd commented:
> `-Zscrape-examples` is an unstable feature; I'd suggest reaching out to the developers of `-Zscrape-examples` to work out what to do here.

I've reached out to @willcrichton about #10500, and turns out they had already fixed it amongst some of the changes in https://github.com/rust-lang/cargo/pull/10343. 

#10343 may involve more coordination with rustc/rustdoc, waiting for stabilization, etc., compared to this PR so I've incorporated the minimum change to fix the issue here (with proper attribution) in case it's merged first, and added a few of the #10500 cases as tests.

With that fixed, and CI passing, this looks ready for review.

ehuss commented:
Can you split the docscrape change into a separate PR?  I would prefer to track and review it separately.

lqd commented:
@ehuss sure. I've split the commits into #10524.

lqd commented:
Now that #10533 has landed, CI passes on this PR.

ehuss commented:
OK, sorry, I finally got around to posting the data:

https://docs.google.com/spreadsheets/d/1lru9ibjHLaXdFbROFIZXcvzXBKDDewGXjRYuvwJ0Xp8/edit?usp=sharing

To read this:

* "Units Diff" - This is the number of extra `rustc` invocations.
* "jxx" - This is the concurrency used.
* "mean Factor" - This is the average factor of the wall-clock of several runs comparing before and after the PR. Example, 0.5 means this PR runs twice as fast as before the PR.
* "system Factor" - The factor of the user-space time comparing before and after.
* "user Factor" - The factor of the system-space time comparing before and after.

There's two sets of data. The one up top is from a Linux system was 16/32 CPUs. The bottom one is from a macOS system with 6/12 CPUs (it has fewer projects since I didn't want to wait for the rest).

There are little arrows in the headers where you can click to show hidden columns that contain the raw data.

This includes some projects that were intentionally selected to have skewed results (those that have a high count of shared dependencies) in order to examine some potentially poor scenarios. All of them should have some proc-macros or build scripts. I find it difficult to select "real world" projects, since they tend to be difficult to find, get access to, build correctly, or take too long to build.

@rust-lang/cargo Does anyone have any thoughts or conclusions you want to draw from this information? I'm still on the fence. It does look like wall clock time is improved in many situations, with a few situations made significantly worse like `cargo-crev`. 

I feel like opportunistically sharing dependencies would be better, but I also feel like Cargo's "share things if possible" logic is already quite complex, and adding more logic to it would make it worse.

epage commented:
If I'm reading this correctly, it looks like the relative improvement is small except for
- most `cargo check` runs
- `toml-rs` and `gluon`s `test --no-run` runs

I'm assuming `cargo check` made a big difference because `build` runs have the builds dominate, taking up a larger percentage.  At first, I wondered about special profiles just for `cargo check` to make it even faster because of how often it is run but fresh-build check times don't matter as much as repeat runs.

I'm curious why we saw a difference with `test --no-run` in those two cases.  We build more targets which includes more linking which I assume would have dominated things even more than `build` does over `check`, making the relative gains even smaller.  So why is it doing better?   **Oh, are test targets built "for the host", so we aren't putting in debug info which saves us on link time which is dominating due to a lot of small test binaries?  Is this a bug with the implementation as I assume people will be debugging their tests and want debug info?**  This would instead be helped if we explored automatic collection into a single test binary, mimicing what `cargo` does with its `tests/testsuite`.

epage commented:
> For toml-rs, for example, the reason you're seeing a noticeable improvement is that many of the proc-macros dependencies are enabled for tests only, and the total number of crates built with debuginfo is reduced from 20 to 10 (all the popular proc-macros libraries and their build scripts

Ah, I had missed that most of those dependencies were introduced exclusively for testing, thanks for finding the actual root cause!

> So the question looks to be more about which is the better default, to fall into the "pit of success" as we know users will rarely change defaults.

I agree.  I feel we should prioritize individuals, local and CI.  Large projects or companies will devote more resources to optimizing things.

lqd commented:
Closing this PR in favor of https://github.com/rust-lang/cargo/pull/11252.
